File: tests/test_equivalence.py
"""Cross-seed / cross-backend equivalence tests."""

from __future__ import annotations

import numpy as np
import pytest

from magical import estimation


def _correlation(a: np.ndarray, b: np.ndarray) -> float:
    """Pearson correlation between two flattened arrays.

    Falls back to 1.0 when both flat vectors are constant (degenerate case
    where every entry is identical — correlation is undefined but the arrays
    agree perfectly).
    """
    x = np.asarray(a).ravel()
    y = np.asarray(b).ravel()
    if np.std(x) == 0 and np.std(y) == 0:
        return 1.0
    return float(np.corrcoef(x, y)[0, 1])


def test_numpy_vs_numpy_cross_seed_high_correlation(fixture) -> None:
    """Two independent numpy runs at different seeds should agree well.

    We use a small iteration count so tests stay fast; the correlation is
    a smoke check that the sampler is exploring roughly the same posterior.
    """
    n_iter = 20
    a = estimation(
        fixture.loaded_data, fixture.candidate_circuits, fixture.initial_model,
        iteration_num=n_iter, backend="numpy", seed=1,
    )
    b = estimation(
        fixture.loaded_data, fixture.candidate_circuits, fixture.initial_model,
        iteration_num=n_iter, backend="numpy", seed=2,
    )
    corr_b = _correlation(a.TF_Peak_Binding_prob, b.TF_Peak_Binding_prob)
    corr_l = _correlation(a.Peak_Gene_Looping_prob, b.Peak_Gene_Looping_prob)
    assert corr_b >= 0.90, f"TF-Peak correlation too low: {corr_b:.3f}"
    assert corr_l >= 0.90, f"Peak-Gene correlation too low: {corr_l:.3f}"


def test_numpy_vs_cpp_placeholder(fixture) -> None:
    """Will be filled in during stage P2 (pybind11 backend)."""
    pytest.importorskip("magical._cpp", reason="cpp backend not implemented yet (stage P2)")
    pytest.skip("filled in during stage P2")

File: re-implementation/tests/testthat/test-equivalence.R
# Cross-backend statistical equivalence.
#
# Under Option B (Armadillo RNG for the C++ backend) the two backends will
# NOT produce identical draws, but pooled posterior probabilities from
# adequately-mixed chains should agree closely. These tests define what
# "closely" means as a contract.
#
# During stage 1 the "cpp" backend is not implemented, so the cpp-vs-r test
# is skipped. The r-vs-r stability test still runs and guards against
# nondeterminism creeping into the R implementation.

context("Cross-backend equivalence")

# Threshold for pooled-posterior agreement on the demo-scale problem.
# On tiny synthetic fixtures with few iterations, chain-to-chain noise
# dominates, so we test a weaker property: high correlation of the pooled
# means from two independent runs of the SAME backend.
CORR_TOL_SAME_BACKEND    <- 0.90   # r-vs-r, different seeds, tiny data
CORR_TOL_ACROSS_BACKENDS <- 0.90   # cpp-vs-r, matched seed intent

run_backend <- function(fx, backend, seed, n_iter) {
  set.seed(seed)
  MAGICAL_estimation(fx$loaded_data, fx$Candidate_circuits, fx$Initial_model,
                     iteration_num = n_iter, backend = backend)
}

matrix_corr <- function(a, b) {
  a_vec <- as.numeric(as.matrix(a))
  b_vec <- as.numeric(as.matrix(b))
  # Constant vectors -> undefined correlation; treat as perfect agreement iff equal.
  if (sd(a_vec) == 0 || sd(b_vec) == 0) {
    return(if (isTRUE(all.equal(a_vec, b_vec))) 1 else NA_real_)
  }
  cor(a_vec, b_vec)
}

test_that("two R-backend runs with different seeds yield correlated posteriors", {
  fx <- make_synthetic_fixture(seed = 42L)

  run_a <- run_backend(fx, "r", seed = 1001L, n_iter = 200L)
  run_b <- run_backend(fx, "r", seed = 2002L, n_iter = 200L)

  binding_corr <- matrix_corr(run_a$TF_Peak_Binding_prob,
                              run_b$TF_Peak_Binding_prob)
  looping_corr <- matrix_corr(run_a$Peak_Gene_Looping_prob,
                              run_b$Peak_Gene_Looping_prob)

  expect_true(!is.na(binding_corr) && binding_corr >= CORR_TOL_SAME_BACKEND,
              info = sprintf("binding corr = %.3f", binding_corr))
  expect_true(!is.na(looping_corr) && looping_corr >= CORR_TOL_SAME_BACKEND,
              info = sprintf("looping corr = %.3f", looping_corr))
})

test_that("cpp backend agrees with R backend up to chain-to-chain noise", {
  # Compilation gated behind MAGICAL_TEST_CPP_BUILD=1 (same flag as
  # test-cpp-build.R) so a plain `testthat.R` run stays fast and doesn't
  # invoke the C++ toolchain.
  skip_if(Sys.getenv("MAGICAL_TEST_CPP_BUILD") != "1",
          "set MAGICAL_TEST_CPP_BUILD=1 to enable")
  skip_if_not_installed("Rcpp")
  skip_if_not_installed("RcppArmadillo")

  ok <- tryCatch({ magical_load_cpp(); TRUE },
                 error = function(e) { message(conditionMessage(e)); FALSE })
  skip_if_not(ok, "magical.cpp failed to compile")
  skip_if_not(exists("magical_estimation_cpp", mode = "function"),
              "cpp entrypoint not exported")

  fx <- make_synthetic_fixture(seed = 7L)

  run_r   <- run_backend(fx, "r",   seed = 5150L, n_iter = 200L)
  run_cpp <- run_backend(fx, "cpp", seed = 5150L, n_iter = 200L)

  binding_corr <- matrix_corr(run_r$TF_Peak_Binding_prob,
                              run_cpp$TF_Peak_Binding_prob)
  looping_corr <- matrix_corr(run_r$Peak_Gene_Looping_prob,
                              run_cpp$Peak_Gene_Looping_prob)

  expect_true(binding_corr >= CORR_TOL_ACROSS_BACKENDS,
              info = sprintf("binding corr = %.3f", binding_corr))
  expect_true(looping_corr >= CORR_TOL_ACROSS_BACKENDS,
              info = sprintf("looping corr = %.3f", looping_corr))
})
